File: web_version/pmugg_files/util/fastMath.js
ms.fastMath = {};

ms.fastMath.mat = function(data) {
	this.data = data;
	this._size = [data.length, data.length > 0 ? data[0].length : 0];
};

ms.fastMath.mat.prototype.size = function() {
	return this._size;
};

ms.fastMath.mat.prototype.set = function(index, value) {
	this.data[index[0]][index[1]] = value;
};

ms.fastMath.mat.prototype.get = function(index, value) {
	return this.data[index[0]][index[1]];
};

ms.fastMath.mat.prototype.valueOf = function() {
	return this.data;
};

ms.fastMath.mat.prototype.subset = function(indices, opt_replacement) {
	return ms.fastMath.subset(this, indices, opt_replacement);
};

ms.fastMath.matrix = function(data) {
	return new ms.fastMath.mat(data);
};

ms.fastMath.zeros = function(size0, size1) {
	var row = new Array(size1).fill(0);
	var rows = [];
	for (var i = 0; i < size0; i++) {
		rows.push(row.slice());
	}
	return new ms.fastMath.mat(rows);
};

// Assumes the matrices are the same size.
ms.fastMath.add = function(A, B) {
	var result = ms.fastMath.zeros(A._size[0], A._size[1]);
	var a = A.data;
	var b = B.data;
	for (var i = 0; i < A._size[0]; i++) {
		for (var j = 0; j < A._size[1]; j++) {
			result.data[i][j] = a[i][j] + b[i][j];
		}
	}
	return result;
};

// Assumes the matrices are the same size.
ms.fastMath.subtract = function(A, B) {
	var result = ms.fastMath.zeros(A._size[0], A._size[1]);
	var a = A.data;
	var b = B.data;
	for (var i = 0; i < A._size[0]; i++) {
		for (var j = 0; j < A._size[1]; j++) {
			result.data[i][j] = a[i][j] - b[i][j];
		}
	}
	return result;
};

ms.fastMath.det2x2 = function(a, b, c, d) {
	return a * d - b * c;
};

ms.fastMath.det = function(A) {
	if (A._size[0] == 2 && A._size[1] == 2) {
		var a = A.data;
		return ms.fastMath.det2x2(a[0][0], a[0][1], a[1][0], a[1][1]);
	} else if (A._size[0] == 3 && A._size[1] == 3) {
		var data = A.data;
		var a = data[0][0];  var b = data[0][1];  var c = data[0][2];
		var d = data[1][0];  var e = data[1][1];  var f = data[1][2];
		var g = data[2][0];  var h = data[2][1];  var i = data[2][2];
		return a*e*i + b*f*g + c*d*h - c*e*g - b*d*i - a*f*h;
	} else {
		ms.alert('det only implemented for 2x2 and 3x3.');
	}
};

ms.fastMath.inv = function(A) {
	if (A._size[0] == 2 && A._size[1] == 2) {
		var det1 = 1.0 / ms.fastMath.det(A);
		var a = A.data;
		var result = [[det1 * a[1][1], -det1 * a[0][1]], [-det1 * a[1][0], det1 * a[0][0]]];
		return ms.fastMath.matrix(result);
	} else if (A._size[0] == 3 && A._size[1] == 3) {
		var det1 = 1.0 / ms.fastMath.det(A);
		var data = A.data;
		var result = [];
		for (var i = 0; i < 3; i++) {
			var row = [];
			for (var j = 0; j < 3; j++) {
				var s = (j + 1) % 3;
				var t = (j + 2) % 3;
				var u = (i + 1) % 3;
				var v = (i + 2) % 3;
				var a = data[s][u];
				var b = data[t][u];
				var c = data[s][v];
				var d = data[t][v];
				row.push(det1 * ms.fastMath.det2x2(a, b, c, d));
			}
			result.push(row);
		}
		return ms.fastMath.matrix(result);
	} else {
		ms.alert('inv only implemented for 2x2 and 3x3.');
	}
};

// Assume these are both column vectors.
ms.fastMath.dot = function(A, B) {
	var a = A.data;
	var b = B.data;
	var sum = 0;
	for (var i = 0; i < A._size[0]; i++) {
		sum += a[i][0] * b[i][0];
	}
	return sum;
};

ms.fastMath.transpose = function(A) {
	var result = ms.fastMath.zeros(A._size[1], A._size[0]);
	var a = A.data;
	for (var i = 0; i < A._size[0]; i++) {
		for (var j = 0; j < A._size[1]; j++) {
			result.data[j][i] = a[i][j];
		}
	}
	return result;
};

ms.fastMath.multiply = function(A, B) {
	if (typeof B == 'number') {
		var result = ms.fastMath.zeros(A._size[0], A._size[1]);
		var a = A.data;
		var r = result.data;
		for (var i = 0; i < A._size[0]; i++) {
			for (var j = 0; j < A._size[1]; j++) {
				r[i][j] = a[i][j] * B;
			}
		}
		return result;
	} else {
		var result = ms.fastMath.zeros(A._size[0], B._size[1]);
		var a = A.data;
		var b = B.data;
		var r = result.data
		for (var i = 0; i < A._size[0]; i++) {
			for (var j = 0; j < B._size[1]; j++) {
				var cij = 0;
				for (var k = 0; k < A._size[1]; k++) {
					cij += a[i][k] * b[k][j];
				}
				r[i][j] = cij;
			}
		}
		return result;
	}
};

ms.fastMath.index = function(index0, index1) {
	return [index0, index1];
};

ms.fastMath.range = function(lower, upper) {
	var result = [];
	for (var i = lower; i < upper; i++) {
		result.push(i);
	}
	return result;
};

ms.fastMath.subset = function(A, indices, opt_replacement) {
	var indices0 = indices[0];
	var indices1 = indices[1];
	var a = A.data;
	if (opt_replacement) {
		var replacement = opt_replacement.data;
		for (var i = 0; i < indices0.length; i++) {
			var index0 = indices0[i];
			for (var j = 0; j < indices1.length; j++) {
				var index1 = indices1[j];
				a[index0][index1] = replacement[i][j];
			}
		}
		return A;
	} else {
		var result = mathG.zeros(indices0.length, indices1.length);
		var r = result.data;
		for (var i = 0; i < indices0.length; i++) {
			var index0 = indices0[i];
			for (var j = 0; j < indices1.length; j++) {
				var index1 = indices1[j];
				r[i][j] = a[index0][index1];
			}
		}
		return result;
	}
};

// This changes A.
ms.fastMath.concat = function(A, B) {
	A._size[1] += B._size[1];
	var size0 = A._size[0];
	var bSize1 = B._size[1];
	var a = A.data;
	var b = B.data;
	for (var i = 0; i < size0; i++) {
		for (var j = 0; j < bSize1; j++) {
			a[i].push(b[i][j]);
		}
	}
	return A;
};

ms.fastMath.test = function() {
	var A = [[0.5111387281709159, -0.2797539798576054], [0.351428224100073, 0.7387129829558294]];
	var B = [[0.9243097344671738, 0.8126861460327792], [0.20270884757161012, 0.4439440454855923]];
	var A0 = math.matrix(A);
	var B0 = math.matrix(B);
	var A1 = ms.fastMath.matrix(A);
	var B1 = ms.fastMath.matrix(B);
	
	var compare = function(result0, result1) {
		console.log(math.subtract(result0, math.matrix(result1.data))._data);
	}
	compare(
		math.add(A0, B0),
		ms.fastMath.add(A1, B1)
	);
	compare(
		math.subtract(A0, B0),
		ms.fastMath.subtract(A1, B1)
	);
	compare(
		math.inv(A0),
		ms.fastMath.inv(A1)
	);
	compare(
		math.multiply(A0, B0),
		ms.fastMath.multiply(A1, B1)
	);
}

File: cpp_version/pmugg/util/fast_math.cpp
#include "fast_math.h"
#include <stdexcept>
#include <cmath>

namespace ms {

FastMath::Matrix::Matrix(const std::vector<std::vector<double>>& d)
    : data(d) {
    size = {static_cast<int>(data.size()),
            static_cast<int>(data.empty() ? 0 : data[0].size())};
}

void FastMath::Matrix::set(const std::vector<int>& index, double value) {
    data[index[0]][index[1]] = value;
}

double FastMath::Matrix::get(const std::vector<int>& index) const {
    return data[index[0]][index[1]];
}

FastMath::Matrix FastMath::Matrix::subset(const std::vector<std::vector<int>>& indices,
                                        const Matrix* replacement) {
    const auto& indices0 = indices[0];
    const auto& indices1 = indices[1];

    if (replacement) {
        for (size_t i = 0; i < indices0.size(); i++) {
            for (size_t j = 0; j < indices1.size(); j++) {
                data[indices0[i]][indices1[j]] = replacement->data[i][j];
            }
        }
        return *this;
    }

    std::vector<std::vector<double>> result(indices0.size(),
                                          std::vector<double>(indices1.size()));
    for (size_t i = 0; i < indices0.size(); i++) {
        for (size_t j = 0; j < indices1.size(); j++) {
            result[i][j] = data[indices0[i]][indices1[j]];
        }
    }
    return Matrix(result);
}

FastMath::Matrix& FastMath::Matrix::concat(const Matrix& B) {
    size[1] += B.size[1];
    for (size_t i = 0; i < data.size(); i++) {
        data[i].insert(data[i].end(), B.data[i].begin(), B.data[i].end());
    }
    return *this;
}

FastMath::Matrix FastMath::matrix(const std::vector<std::vector<double>>& data) {
    return Matrix(data);
}

FastMath::Matrix FastMath::zeros(int size0, int size1) {
    return Matrix(std::vector<std::vector<double>>(size0,
                                                 std::vector<double>(size1, 0.0)));
}

FastMath::Matrix FastMath::add(const Matrix& A, const Matrix& B) {
    auto result = zeros(A.size[0], A.size[1]);
    for (int i = 0; i < A.size[0]; i++) {
        for (int j = 0; j < A.size[1]; j++) {
            result.data[i][j] = A.data[i][j] + B.data[i][j];
        }
    }
    return result;
}

FastMath::Matrix FastMath::subtract(const Matrix& A, const Matrix& B) {
    auto result = zeros(A.size[0], A.size[1]);
    for (int i = 0; i < A.size[0]; i++) {
        for (int j = 0; j < A.size[1]; j++) {
            result.data[i][j] = A.data[i][j] - B.data[i][j];
        }
    }
    return result;
}

FastMath::Matrix FastMath::multiply(const Matrix& A, const Matrix& B) {
    auto result = zeros(A.size[0], B.size[1]);
    for (int i = 0; i < A.size[0]; i++) {
        for (int j = 0; j < B.size[1]; j++) {
            double sum = 0;
            for (int k = 0; k < A.size[1]; k++) {
                sum += A.data[i][k] * B.data[k][j];
            }
            result.data[i][j] = sum;
        }
    }
    return result;
}

FastMath::Matrix FastMath::multiply(const Matrix& A, double scalar) {
    auto result = zeros(A.size[0], A.size[1]);
    for (int i = 0; i < A.size[0]; i++) {
        for (int j = 0; j < A.size[1]; j++) {
            result.data[i][j] = A.data[i][j] * scalar;
        }
    }
    return result;
}

FastMath::Matrix FastMath::transpose(const Matrix& A) {
    auto result = zeros(A.size[1], A.size[0]);
    for (int i = 0; i < A.size[0]; i++) {
        for (int j = 0; j < A.size[1]; j++) {
            result.data[j][i] = A.data[i][j];
        }
    }
    return result;
}

double FastMath::det2x2(double a, double b, double c, double d) {
    return a * d - b * c;
}

double FastMath::det(const Matrix& A) {
    if (A.size[0] == 2 && A.size[1] == 2) {
        return det2x2(A.data[0][0], A.data[0][1],
                     A.data[1][0], A.data[1][1]);
    } else if (A.size[0] == 3 && A.size[1] == 3) {
        const auto& d = A.data;
        double a = d[0][0], b = d[0][1], c = d[0][2];
        double e = d[1][0], f = d[1][1], g = d[1][2];
        double h = d[2][0], i = d[2][1], j = d[2][2];
        return a*f*j + b*g*h + c*e*i - c*f*h - b*e*j - a*g*i;
    }
    throw std::runtime_error("det only implemented for 2x2 and 3x3");
}

FastMath::Matrix FastMath::inv(const Matrix& A) {
    if (A.size[0] == 2 && A.size[1] == 2) {
        double det1 = 1.0 / det(A);
        const auto& a = A.data;
        return matrix({{det1 * a[1][1], -det1 * a[0][1]},
                      {-det1 * a[1][0], det1 * a[0][0]}});
    } else if (A.size[0] == 3 && A.size[1] == 3) {
        double det1 = 1.0 / det(A);
        const auto& d = A.data;
        std::vector<std::vector<double>> result(3, std::vector<double>(3));
        
        for (int i = 0; i < 3; i++) {
            for (int j = 0; j < 3; j++) {
                int s = (j + 1) % 3;
                int t = (j + 2) % 3;
                int u = (i + 1) % 3;
                int v = (i + 2) % 3;
                double a = d[s][u];
                double b = d[t][u];
                double c = d[s][v];
                double e = d[t][v];
                result[i][j] = det1 * det2x2(a, b, c, e);
            }
        }
        return matrix(result);
    }
    throw std::runtime_error("inv only implemented for 2x2 and 3x3");
}

double FastMath::dot(const Matrix& A, const Matrix& B) {
    double sum = 0;
    for (int i = 0; i < A.size[0]; i++) {
        sum += A.data[i][0] * B.data[i][0];
    }
    return sum;
}

std::vector<int> FastMath::index(int index0, int index1) {
    return {index0, index1};
}

std::vector<int> FastMath::range(int lower, int upper) {
    std::vector<int> result;
    for (int i = lower; i < upper; i++) {
        result.push_back(i);
    }
    return result;
}

void FastMath::test() {
    // Test matrix operations
    Matrix A({{0.5111, -0.2797}, {0.3514, 0.7387}});
    Matrix B({{0.9243, 0.8126}, {0.2027, 0.4439}});

    auto sum = add(A, B);
    auto diff = subtract(A, B);
    auto prod = multiply(A, B);
    auto invA = inv(A);

    // Add verification code as needed
}

} // namespace ms 